### PR TITLE
[TH2-3589] Flags for first and last messages

### PR DESCRIPTION
**Pull request by Toru8P**

### CODE REVIEW

File: .github/workflows/dev-java-publish-sonatype.yml [hunk 1]
```diff
@@ -24,7 +24,7 @@ jobs:
           property: release_version
       - name: Build custom release version
         id: release_ver
-        run: echo ::set-output name=value::"${{ steps.ver.outputs.value }}-${{ steps.branch.outputs.branch_name }}-SNAPSHOT"
+        run: echo ::set-output name=value::"${{ steps.ver.outputs.value }}-${{ steps.branch.outputs.branch_name }}-${{ github.run_id }}-SNAPSHOT"
```
Comment (OptimumCode) [inline .github/workflows/dev-java-publish-sonatype.yml]: Why do you need this ID for snapshots?

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt [hunk 2]
```diff
@@ -650,6 +668,8 @@ abstract class AbstractFileReader<T : AutoCloseable>(
             get() = Files.exists(path)
         var truncated: Boolean = false
             private set
+        var readState: ReadState = ReadState.START
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt:670]: I think we should not expose the state setter (because the FileHolder is not a private class). We can encapsulate the state changes inside some methods

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt [hunk 3]
```diff
@@ -771,6 +795,11 @@ abstract class AbstractFileReader<T : AutoCloseable>(
     companion object {
         private val LOGGER = KotlinLogging.logger { }
 
+        const val MESSAGE_STATUS_PROPERTY = "STATUS"
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: I think we should keep properties names more unique to avoid possible collisions
```suggestion
        const val MESSAGE_STATUS_PROPERTY = "th2.read.order_marker"
```

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt [hunk 4]
```diff
@@ -771,6 +795,11 @@ abstract class AbstractFileReader<T : AutoCloseable>(
...
+        const val MESSAGE_STATUS_SINGLE = "SINGLE"
+        const val MESSAGE_STATUS_FIRST = "START"
+        const val MESSAGE_STATUS_LAST = "FIN"
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: Let's make them lower case. It is more readable in my oppinion
```suggestion
        const val MESSAGE_STATUS_SINGLE = "single"
        const val MESSAGE_STATUS_FIRST = "start"
        const val MESSAGE_STATUS_LAST = "fin"
```

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt [hunk 5]
```diff
@@ -458,10 +472,14 @@ abstract class AbstractFileReader<T : AutoCloseable>(
...
                     if (content.isNotEmpty()) {
                         LOGGER.trace { "Read ${content.size} message(s) for $streamId from ${holder.path}" }
+                        holder.readState = FileHolder.ReadState.IN_PROGRESS
                         break
                     }
                 }
             } while (canParse && hasMoreData)
+
+            holder.readState = if (!hasMoreData) FileHolder.ReadState.FIN else FileHolder.ReadState.IN_PROGRESS
+            LOGGER.trace { "After read holder had status: ${holder.readState}" }
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: I think we should move that to `processUpdates` method to keep all changes to the state close to each other

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt [hunk 6]
```diff
@@ -202,6 +204,13 @@ abstract class AbstractFileReader<T : AutoCloseable>(
                     }
 
                     finalContent.also { content ->
+                        if (content.size == 1 && lastState == FileHolder.ReadState.START && fileHolder.readState == FileHolder.ReadState.FIN) {
+                            content.first().markSingle()
+                        } else {
+                            if (lastState == FileHolder.ReadState.START) content.first().markFirst()
+                            if (fileHolder.readState == FileHolder.ReadState.FIN) content.last().markLast()
+                        }
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: I think it looks better as a `when` statement
```suggestion
                        when {
                          content.size == 1 && lastState == FileHolder.ReadState.START && fileHolder.readState == FileHolder.ReadState.FIN ->
                            content.first().markSingle()
                          lastState == FileHolder.ReadState.START -> content.first().markFirst()
                          fileHolder.readState == FileHolder.ReadState.FIN -> content.last().markLast()
                        }
```

File: src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReader.kt [hunk 7]
```diff
@@ -131,7 +167,9 @@ internal class TestAbstractFileReader : AbstractReaderTest() {
     }
 
     @Test
+    @Disabled //FIXME
```
Comment (OptimumCode) [inline src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReader.kt]: What is the problem with the test?

**Review by OptimumCode — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt (hunk 6 above)
Comment (Toru8P) [reply src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: In your example chain when two statments can be true isnt allowed, one group can have first and also last states

File: .github/workflows/dev-java-publish-sonatype.yml (hunk 1 above)
Comment (Toru8P) [reply .github/workflows/dev-java-publish-sonatype.yml]: Easier to understand which version was used at some particular moment while developing

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt (hunk 6 above)
Comment (OptimumCode) [reply src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: Yes, you are right. My mistake

File: .github/workflows/dev-java-publish-sonatype.yml (hunk 1 above)
Comment (OptimumCode) [reply .github/workflows/dev-java-publish-sonatype.yml]: But it completely breaks the idea of snapshots. Each time we have a new version and we need to update it

File: README.md [hunk 8]
```diff
@@ -117,6 +117,10 @@ The module you need is `com.fasterxml.jackson.datatype:jackson-datatype-jsr310`.
 
 ## Changes
 
+### 1.3.0
+
++ Adds new flags for first `START` and last `FIN` messages
```
Comment (OptimumCode) [inline README.md]: Please, correct flags's names in the readme. Also, please add a property name to the readme file (probably to the change log and to the description block)

File: src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt [hunk 9]
```diff
@@ -462,6 +479,7 @@ abstract class AbstractFileReader<T : AutoCloseable>(
                     }
                 }
             } while (canParse && hasMoreData)
+            holder.updateState()
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/read/file/common/AbstractFileReader.kt]: I still think that we should move it to the `processUpdates` method. It is close to all other state changes

File: src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReader.kt [hunk 10]
```diff
@@ -132,6 +168,7 @@ internal class TestAbstractFileReader : AbstractReaderTest() {
 
     @Test
     internal fun `reads data from log rotation pattern`() {
+
```
Comment (OptimumCode) [inline src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReader.kt]: Please, remove the changes

**Review by OptimumCode — CHANGES_REQUESTED** [review]

File: .github/workflows/dev-java-publish-sonatype.yml (hunk 1 above)
Comment (Toru8P) [reply .github/workflows/dev-java-publish-sonatype.yml]: `Maven checks for a new SNAPSHOT version in a remote repository for **every build**.`

Okay, got it, thanks for pointing this out

File: src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReaderSustained.kt [hunk 11]
```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2020-2021 Exactpro (Exactpro Systems Limited)
```
Comment (OptimumCode) [inline src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReaderSustained.kt]: Please, correct the year in the copyright. Also, why did you move the tests to another class

File: README.md [hunk 12]
```diff
@@ -115,8 +115,20 @@ You need to use [JavaTime modules](https://github.com/FasterXML/jackson-modules-
 
 The module you need is `com.fasterxml.jackson.datatype:jackson-datatype-jsr310`.
 
+#### Behavior details
+
+Each file will produce group of messages that can include `th2.read.order_marker` property:
+ * First and last messages will be marked as `start` and `fin`. If first message wasn't sent, `start` mark will be moved to next one 
+until some messages will be actually sent.
+ * If file produced only one message, it will be marked as `single`
```
Comment (OptimumCode) [inline README.md]: Please, add information about the `leaveLastFileOpen` option. The property won't be set in this case

Comment (OptimumCode) [reply README.md]: Did you have a chance to look at it?

File: src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReaderSustained.kt (hunk 11 above)
Comment (Toru8P) [reply src/test/kotlin/com/exactpro/th2/read/file/common/TestAbstractFileReaderSustained.kt]: i did not move tests, i renamed it due continues behavior, tests for single read was created with last name

File: README.md [hunk 13]
```diff
@@ -120,8 +120,9 @@ The module you need is `com.fasterxml.jackson.datatype:jackson-datatype-jsr310`.
 Each file will produce group of messages that can include `th2.read.order_marker` property:
  * First and last messages will be marked as `start` and `fin`. If first message wasn't sent, `start` mark will be moved to next one 
 until some messages will be actually sent.
- * If file produced only one message, it will be marked as `single`
+ * If file produced only one message, it will be marked as `single` 
 
+ **NOTE**: If option `leaveLastFileOpen` was turned on means no flags will be generated, this option provides continues data without certain end
```
Comment (OptimumCode) [inline README.md]:
```suggestion
 **NOTE**: If option `leaveLastFileOpen` is turned on means no flags will be generated because in this case we cannot determinate the last message read from the file
```

**Review by OptimumCode — APPROVED** [review]
